### ISSUE TITLE
Golang dependencies on generated code

### ISSUE THREAD

**Issue by ronnylt** [renovatebot/renovate#4490]
**What Renovate type are you using?**
The hosted App

**Describe the bug**
When having a Golang codebase with dependencies on generated code (i.e. Protocol Buffers) a previous step is required to generate the code before running `go mod ...` . If the generated code is not present, the `go mod ...` will fail.

**To Reproduce**
Any Golang codebase with dependencies on generated code, for example a ' go generate` invocation required before running `go mod ...`

**Expected behavior**
Modules updated.

**Screenshots**

```

Command failed: docker run --rm -v /mnt/renovate/gh/xxxx/xxxxx:/mnt/renovate/gh/xxxx/xxxx -v /tmp/renovate-cache/others/go:/tmp/renovate-cache/others/go -e GOPATH -e GOPROXY -e CGO_ENABLED=0 -w /mnt/renovate/gh/xxx/xxx renovate/go bash -c "git config --global url.\"https://x-access-token:**redacted**@github.com/\".insteadOf \"https://github.com/\" && go get ./..."
build xxx/cmd/xxx cannot load github.com/xxx/protos/go/xxxx: no matching versions for query "latest"


```

**Comment by rarkins** [comment]
Will `go get` still succeed? Or it will also fail?

**Comment by diagonalfish** [comment]
I can confirm this problem in go 1.13 - 'go get' or any other 'go' command that acts on the module will fail due to the missing generated code with the error in the OP.

I was considering submitting a feature request for "preUpdateActions" that'd be done before attempting to do updates, which would allow you to work around sort of problem.

**Comment by xens** [comment]
I'm wondering if there's any progress on that one ? I'd love to help with a PR but I've never contributed to Renovate so far, so any tips / guidance is welcome.

**Comment by rarkins** [comment]
This is still in the "Needs Requirements" category, meaning we need to work out how we'd do it. We already have the concept of [postUpdateTasks](https://docs.renovatebot.com/configuration-options/#postupgradetasks) so the suggestion of preUpdateTasks seems logical. This would be applicable for *self hosted* only so that you have full flexibility of what you need to run.

**Comment by xens** [comment]
@rarkins so if I understand correctly the concept of preUpdateTasks would be a kind of workaround as it'd allows us to pre-execute the missing `go generate` command before Renovate issue the `go mod` one. Do we also want to work on fixing the logic inside the golang manager ?

**Comment by rarkins** [comment]
No, I think we have a misunderstanding. At this point it's *intended* that we don't allow arbitrary code execution by default, such as `go generate` would allow. I'm suggesting that you can *add a feature* to allow arbitrary code execution prior to updating in case you are running self-hosted and don't mind the risk.

By the way, the original description in this issue seems a little wrong because it says `go generate` is required before any `go mod` however as per the log we are actually running `go get` and not any `go mod` command. `go mod` is only run if people opt into go mod tidy.

**Comment by xens** [comment]
Ok gotcha thanks for the explanation that makes a lot of sense.

**Comment by rarkins** [comment]
Right now we have:
- `postUpdateOptions`: simple, safe opt-in things to be run after updating, if opted into
- `postUpdateTasks`: similar to the above, but can be completely arbitrary code
- `allowedPostUpgradeCommands`: something which lets bot owners have a whitelist of allowed commands/patterns

We could perhaps rename and reuse `allowedPostUpgradeCommands` e.g. into `allowedUpgradeCommands` to be shared between `preUpdateTasks` and `postUpdateTasks`, as there doesn't seem to be any reason to differentiate between them from a security point of view.

**Comment by xens** [comment]
> postUpdateTasks: similar to the above, but can be completely arbitrary code

you mean `postUpgradeTasks` I guess, but yeah your proposal sound good !

**Comment by rarkins** [comment]
Thanks, corrected

**Comment by vitkep** [comment]
Any progress on this issue?

We have a problem with renovate and gererated modules. We have a module, which is referenced by an import in code but will generated on build time with make. So if renovate tries to go throw imports with `go get` it cannot resolve the import. There is any solution on that? Can I execute `make build` before renovate fetching dependencies? This problem should have many projects working with module generation.
